zaza/openstack/charm_tests/heat/tests.py

```diff
@@ -0,0 +1,355 @@
+                                                _server_id,
+                                                expected_status="ACTIVE",
+                                                msg="nova instance")
+        logging.info('Nova instance reached ACTIVE status')
+
+    def _stack_delete(self):
+        """Delete a heat stack, verify."""
+        logging.info('Deleting heat stack...')
+        openstack_utils.delete_resource(self.heat_client.stacks,
+                                        STACK_NAME, msg="heat stack")
+
+    def _image_delete(self):
+        """Delete that image."""
+        logging.info('Deleting glance image...')
+        image = self.nova_client.glance.find_image(IMAGE_NAME)
+        openstack_utils.delete_resource(self.glance_client.images,
+                                        image.id, msg="glance image")
+
+    def _keypair_delete(self):
+        """Delete that keypair."""
+        logging.info('Deleting keypair...')
+        openstack_utils.delete_resource(self.nova_client.keypairs,
+                                        KEYPAIR_NAME, msg="nova keypair")
+
+    def test_100_domain_setup(self):
+        """Run required action for a working Heat unit."""
+        # Action is REQUIRED to run for a functioning heat deployment
+        logging.info('Running domain-setup action on heat unit...')
+        unit = zaza.model.get_units(self.application_name)[0]
+        assert unit.workload_status == "active"
+        zaza.model.run_action(unit.entity_id, "domain-setup")
+        zaza.model.block_until_unit_wl_status(unit.entity_id, "active")
+        unit = zaza.model.get_unit_from_name(unit.entity_id)
+        assert unit.workload_status == "active"
+
+    def test_400_heat_resource_types_list(self):
+        """Check default resource list behavior and confirm functionality."""
+        logging.info('Checking default heat resource list...')
+        try:
+            types = list(self.heat_client.resource_types.list())
+            if type(types) is list:
+                logging.info('Resource type list check is ok.')
+            else:
+                msg = 'Resource type list is not a list!'
+                logging.error('{}'.format(msg))
+                raise
+            if len(types) > 0:
+                logging.info('Resource type list is populated '
+                             '({}, ok).'.format(len(types)))
+            else:
+                msg = 'Resource type list length is zero!'
+                logging.error(msg)
+                raise
+        except Exception as e:
+            msg = 'Resource type list failed: {}'.format(e)
+            logging.error(msg)
+            raise
+
+    def test_402_heat_stack_list(self):
+        """Check default heat stack list behavior, confirm functionality."""
+        logging.info('Checking default heat stack list...')
+        try:
+            stacks = list(self.heat_client.stacks.list())
+            if type(stacks) is list:
```

Comment:
Again, `stacks` is always a list if the code gets this far.

Reply:
@ajkavanagh Upon closer inspection, I am going to remove the 402() function since the heat stack is empty at this point, and I do check the stack list in _stack_create() after a stack is created.